keep/step/step.py

```diff
@@ -1,51 +1,198 @@
+                condition.condition_type,
+                condition_compare_to,
+                condition_compare_value,
+                condition_result,
+                condition_alias=condition.condition_alias,
+                **condition.condition_context,
+            )
+
+        # Second, decide if need to run
+        # after all conditions are applied, check if we need to run
+        # there are 2 cases:
+        # 1. a "if" block is supplied, then use it
+        # 2. no "if" block is supplied, then use the AND between all conditions
+        if self.config.get("if"):
+            if_conf = self.config.get("if")
+        else:
+            # create a string of all conditions, separated by "and"
+            if_conf = " and ".join(
+                [f"{{{{ {condition.condition_alias} }}}} " for condition in conditions]
+            )
+
+        # Now check it
+        if if_conf:
+            if_met = self.io_handler.render(if_conf)
+            # Evaluate the condition string
+            if_met = eval(if_met)
+        else:
+            if_met = True
+
+        if not if_met:
+            self.logger.info(
+                "Action %s evaluated NOT to run, Reason: %s evaluated to false.",
+                self.config.get("name"),
+                if_conf,
+            )
+            return
 
+        self.logger.info("Action %s evaluated to run!", self.config.get("name"))
 
+        # Third, check throttling
+        # Now check if throttling is enabled
+        throttled = self._check_throttling(self.config.get("name"))
+        if throttled:
+            self.logger.info("Action %s is throttled", self.config.get("name"))
+            return
+
+        # Last, run the action
+        rendered_value = self.io_handler.render_context(self.provider_context)
+        # if the provider is async, run it in a new event loop
+        if inspect.iscoroutinefunction(self.provider.notify):
+            result = self._run_single_async()
+        # else, just run the provider
+        else:
+            if self.step_type == StepType.STEP:
+                try:
+                    rendered_providers_parameters = {}
+                    for parameter in self.provider_parameters:
+                        rendered_providers_parameters[parameter] = self.io_handler.render(
+                            self.provider_parameters[parameter]
+                        )
+                    step_output =  self.provider.query(**rendered_value)
+                    extra_context = self.provider.expose()
+                    rendered_providers_parameters.update(extra_context)
+                    self.context_manager.set_step_provider_paremeters(
+                        self.step_id, rendered_providers_parameters
+                    )
+                except Exception as e:
+                    raise StepError(e)
+                
+                return step_output
+            self.provider.notify(**rendered_value)
+            return True
```

Comment:
I think we need to try and keep logic as similar as possible, wdyt about the following:

```suggestion
        if inspect.iscoroutinefunction(self.provider.notify) or inspect.iscoroutinefunction(self.provider.query):
            result = self._run_single_async()
        # else, just run the provider
        else:
              try:
                  rendered_providers_parameters = {}
                  for parameter in self.provider_parameters:
                      rendered_providers_parameters[parameter] = self.io_handler.render(
                          self.provider_parameters[parameter]
                      )
                  step_output =  self.provider.query(**rendered_value) if self.step_type == StepType.STEP else self.provider.notify(**rendered_value)
                  extra_context = self.provider.expose()
                  rendered_providers_parameters.update(extra_context)
                  self.context_manager.set_step_provider_paremeters(
                      self.step_id, rendered_providers_parameters
                  )
              except Exception as e:
                  raise StepError(e)
```